### ISSUE TITLE
Evaluation data should be different from training data

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#337]
I think that the evaluation data should not be necessarily the same as the training data.

My idea is that the training data should consist of only those reviews which the algorithm can efficiently deal with. However, the evaluation data should contain all the data that affects future reviewing of any card.

For example, in my opinion,
- ~~All the ratings of a card before a manual reset should be discarded because they are likely to be outliers (and thus, not fit for inclusion in the training data) and they don't affect the future reviewing (and thus, not fit for inclusion in the evaluation data).~~ (already dealt with)
- ~~All the [cram revlogs](https://github.com/open-spaced-repetition/fsrs4anki/commit/3c037cb4068a43df92759f953dcb1783423ee7d5) should be discarded because they are likely to be outliers (and thus, not fit for inclusion in the training data) and they don't affect the future reviewing (and thus, not fit for inclusion in the evaluation data).~~ (already dealt with)
- All the ratings before the `revlog_start_date` should be removed from the training data but they should be included in the evaluation data (because they still affect the future reviewing).
- The outliers discussed in [this issue](https://github.com/open-spaced-repetition/fsrs4anki/issues/275) should be removed from the training data (if you think it is desirable) but they should be included in the evaluation data (because they still affect the future reviewing).

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki/issues/320#issuecomment-1610632305_

**Comment by user1823** [comment]
Additionally, a new variable called something like `evaluation_start_date` should be added.

Reason:
It will allow the user to compare the performance of different optimizations on recent reviews. While optimization over entire history may produce better fitting for old reviews and better overall metrics, it doesn't necessary mean that it will be better for recent reviews. Performance on recent reviews is important because it better represents what would happen when the user will do new reviews, that is the ability of algorithm to make predictions for new data it hasn't been trained on.

_Originally posted by @nb9618 in https://github.com/open-spaced-repetition/fsrs4anki/issues/320#issuecomment-1610933095_

This `evaluation_start_date` variable should be added just before the evaluation begins (i.e., in section 4) so that the user can change it to produce evaluations for different sets of data without needing to train the parameters again.

Also, as I said earlier, the evaluation data should be different from the training data. So, this `evaluation_start_date` should be used to filter the evaluation data and not the training data.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki/issues/320#issuecomment-1610963422_

**Comment by user1823** [comment]
We have left it to the user to decide whether to filter out the suspended cards in the training data or not. But, I think that we should **filter out the suspended cards in the evaluation data**. The reason being that even if the revlogs of the suspended cards can be helpful in training the parameters, those revlogs are of no value in the evaluation process (because those cards are not being reviewed).

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki/issues/270#issuecomment-1612748258_

**Comment by L-M-Sherlock** [comment]
> I think that the evaluation data should not be necessarily the same as the training data.

If the performance in the evaluation data is not good, what it affects the improvement of FSRS or the decision of users? If the evaluation data is the same as the training data, the bad performance could be attributed to the bad structure of algorithm. If they are different, the cause is ambiguous. It may not provide any guiding significance or operability.

**Comment by user1823** [comment]
First of all, the training data and the evaluation data are not completely different. Most of the data is overlapping and only some data is different.

Secondly, the user is only concerned about the performance of the optimized model on the reviews that are actually involved in their review process.

For example, knowing that the model performs well on suspended cards doesn't benefit the user because the user is not actually studying those cards.

For another example, even if the model is not trained on reviews before the `revlog_start_date`, the user still needs to know how the model would perform on those reviews when the user would put the parameters into the scheduler.

**Comment by L-M-Sherlock** [comment]
Recently, it occurred to me that splitting the trainset and testset by time-series is a really good way to evaluate FSRS.

For more details, please see: https://scikit-learn.org/stable/modules/cross_validation.html#time-series-split

**Comment by L-M-Sherlock** [comment]
I create a new benchmark notebook: https://github.com/open-spaced-repetition/fsrs4anki/blob/benchmark/benchmark_tscv.ipynb

**Comment by user1823** [comment]
I am closing this as completed because it is possible to run Optimize and Evaluate on different data in Anki.
